prow/cmd/pipeline/main.go

```diff
@@ -0,0 +1,192 @@
+}
+
+type pipelineConfig struct {
+	client   pipelineset.Interface
+	informer pipelineinfov1alpha1.PipelineRunInformer
+}
+
+// newPipelineConfig returns a client and informer capable of mutating and monitoring the specified config.
+func newPipelineConfig(cfg rest.Config, stop chan struct{}) (*pipelineConfig, error) {
+	bc, err := pipelineset.NewForConfig(&cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure the pipeline CRD is deployed
+	// TODO(fejta): probably a better way to do this
+	if _, err := bc.TektonV1alpha1().PipelineRuns("").List(metav1.ListOptions{Limit: 1}); err != nil {
+		return nil, err
+	}
+
+	// Assume watches receive updates, but resync every 30m in case something wonky happens
+	bif := pipelineinfo.NewSharedInformerFactory(bc, 30*time.Minute)
+	bif.Tekton().V1alpha1().PipelineRuns().Lister()
+	go bif.Start(stop)
+	return &pipelineConfig{
+		client:   bc,
+		informer: bif.Tekton().V1alpha1().PipelineRuns(),
+	}, nil
+}
+
+func main() {
+	o := parseOptions()
+	logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "pipeline"})
+
+	pjutil.ServePProf()
+
+	configAgent := &config.Agent{}
+	if o.config != "" {
+		const ignoreJobConfig = ""
+		if err := configAgent.Start(o.config, ignoreJobConfig); err != nil {
+			logrus.WithError(err).Fatal("failed to load prow config")
+		}
+	}
+
+	configs, err := kube.LoadClusterConfigs(o.kubeconfig, o.pipelineCluster)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error building client configs")
+	}
+
+	if !o.allContexts { // Just the default context please
+		logrus.Warn("Truncating to local and default contexts")
+		configs = map[string]rest.Config{
+			kube.InClusterContext:    configs[kube.InClusterContext],
+			kube.DefaultClusterAlias: configs[kube.DefaultClusterAlias],
+		}
+	}
+	local := configs[kube.InClusterContext]
+
+	stop := stopper()
+
+	kc, err := kubernetes.NewForConfig(&local)
+	if err != nil {
+		logrus.WithError(err).Fatalf("Failed to create local kubernetes client")
+	}
+	pjc, err := prowjobset.NewForConfig(&local)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create prowjob client")
+	}
+	pjif := prowjobinfo.NewSharedInformerFactory(pjc, 30*time.Minute)
+	pjif.Prow().V1().ProwJobs().Lister()
+	go pjif.Start(stop)
+
+	pipelineConfigs := map[string]pipelineConfig{}
+	for context, cfg := range configs {
+		var bc *pipelineConfig
+		bc, err = newPipelineConfig(cfg, stop)
+		if apierrors.IsNotFound(err) {
+			logrus.WithError(err).Warnf("Ignoring %s: knative pipeline CRD not deployed", context)
+			continue
+		}
+		if err != nil {
+			logrus.WithError(err).Fatalf("Failed to create %s pipeline client", context)
+		}
+		pipelineConfigs[context] = *bc
+	}
+
+	controller, err := newController(kc, pjc, pjif.Prow().V1().ProwJobs(), pipelineConfigs, o.totURL, configAgent.Config, kube.RateLimiter(controllerName))
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating controller")
+	}
+
+	if err := controller.Run(1, stop); err != nil {
```

Comment:
Why only 1 worker thread? The build controller uses 2 (also why?).

Reply:
In my understanding, having `2` worker threads ensures that two reconcile loops can run in parallel. This means that 2 items from the queue will be processed in parallel.  It will likely lead to lower latency since the build controller handles 2 types of CRDs (ProwJob and Build). In the same time, can lead to object update conflicts when the subsequent items are of the same type. For instance, one worker has already updated the ProwJob object while the other one is processing the previous version from the cache. Maybe a PATCH update strategy can mitigate this. 

I can also set 2 worker threads for pipeline controller.

Reply:
Set it to 2 worker threads.

Reply:
I don't think multiple versions of the same object are in the queue for the controller at the same time, keys are coalesced IIRC.